### PR TITLE
Add Generator for .copy() Methods

### PR DESCRIPTION
**Pull request by shiki**
This is a continuation of #2238. This adds the ability to generate `.copy()` methods using [Sourcery](https://github.com/krzysztofzablocki/Sourcery). 

👋  Please read p91TBi-2TB-p2 first. 

## Testing

This is just a suggestion. Feel free to diverge. 

1. Make these `class` and `structs` conform to `GeneratedCopiable`: 
    - [`ProductSettings`](https://github.com/woocommerce/woocommerce-ios/blob/develop/Yosemite/Yosemite/Model/Products/ProductSettings.swift)
    - [`Notice`](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/Tools/Notices/Notice.swift)
    - [`Product`](https://github.com/woocommerce/woocommerce-ios/blob/develop/Networking/Networking/Model/Product/Product.swift)
2.  Build the project to make sure there are no errors. 
3. In the terminal, run `rake generate`
4. Confirm that there are no errors. 
5. There should be 2 (generated) files that are not added to the project yet. Add them now. 
    - `WooCommerce/Classes/Copiable/Models+Copiable.generated.swift`
    - `Yosemite/Yosemite/Model/Copiable/Models+Copiable.generated.swift`
6. Build the project
7. Play around with the generated `.copy()` methods. For example, try replacing the code in [`ProductUpdater`](https://github.com/woocommerce/woocommerce-ios/blob/develop/Yosemite/Yosemite/Model/Updaters/ProductUpdater.swift) with `copy()` calls like this:

    ```diff
    diff --git a/Yosemite/Yosemite/Model/Updaters/ProductUpdater.swift b/Yosemite/Yosemite/Model/Updaters/ProductUpdater.swift
    index 4d778e7be..511d5342a 100644
    --- a/Yosemite/Yosemite/Model/Updaters/ProductUpdater.swift
    +++ b/Yosemite/Yosemite/Model/Updaters/ProductUpdater.swift
    @@ -23,66 +23,7 @@ public protocol ProductUpdater {

    extension Product: ProductUpdater {
        public func nameUpdated(name: String) -> Product {
    -        return Product(siteID: siteID,
    -                       productID: productID,
    -                       name: name,
    -                       slug: slug,
    -                       permalink: permalink,
    -                       dateCreated: dateCreated,
    -                       dateModified: dateModified,
    -                       dateOnSaleStart: dateOnSaleStart,
    -                       dateOnSaleEnd: dateOnSaleEnd,
    -                       productTypeKey: productTypeKey,
    -                       statusKey: statusKey,
    -                       featured: featured,
    -                       catalogVisibilityKey: catalogVisibilityKey,
    -                       fullDescription: fullDescription,
    -                       briefDescription: briefDescription,
    -                       sku: sku,
    -                       price: price,
    -                       regularPrice: regularPrice,
    -                       salePrice: salePrice,
    -                       onSale: onSale,
    -                       purchasable: purchasable,
    -                       totalSales: totalSales,
    -                       virtual: virtual,
    -                       downloadable: downloadable,
    -                       downloads: downloads,
    -                       downloadLimit: downloadLimit,
    -                       downloadExpiry: downloadExpiry,
    -                       externalURL: externalURL,
    -                       taxStatusKey: taxStatusKey,
    -                       taxClass: taxClass,
    -                       manageStock: manageStock,
    -                       stockQuantity: stockQuantity,
    -                       stockStatusKey: stockStatusKey,
    -                       backordersKey: backordersKey,
    -                       backordersAllowed: backordersAllowed,
    -                       backordered: backordered,
    -                       soldIndividually: soldIndividually,
    -                       weight: weight,
    -                       dimensions: dimensions,
    -                       shippingRequired: shippingRequired,
    -                       shippingTaxable: shippingTaxable,
    -                       shippingClass: shippingClass,
    -                       shippingClassID: shippingClassID,
    -                       productShippingClass: productShippingClass,
    -                       reviewsAllowed: reviewsAllowed,
    -                       averageRating: averageRating,
    -                       ratingCount: ratingCount,
    -                       relatedIDs: relatedIDs,
    -                       upsellIDs: upsellIDs,
    -                       crossSellIDs: crossSellIDs,
    -                       parentID: parentID,
    -                       purchaseNote: purchaseNote,
    -                       categories: categories,
    -                       tags: tags,
    -                       images: images,
    -                       attributes: attributes,
    -                       defaultAttributes: defaultAttributes,
    -                       variations: variations,
    -                       groupedProducts: groupedProducts,
    -                       menuOrder: menuOrder)
    +        copy(name: name)
        }
    ```

    Does it build? Does the `Product_UpdaterTestCases` still pass?

## Reviewing 

At least 2 reviewers would be awesome. 

## Submitter Checklist

- [x] If it's feasible, I have added unit tests. 
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -102,6 +102,8 @@ def networking_pods
   pod 'CocoaLumberjack', '~> 3.5'
   pod 'CocoaLumberjack/Swift', '~> 3.5'
 
+  pod 'Sourcery', '~> 0.18'
```
Comment (astralbodies) [inline Podfile:105]: Can we limit this pod to only be included in debug configurations? i.e. `, :configurations => ['Debug']`

File: CodeGeneration/Models+Copiable.swifttemplate [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        let name: String
+        /// String, Int, etc 
+        let typeName: String
+        /// NullableCopiableProp or CopiableProp
+        let copiablePropTypeName: String
+        /// If this is not the last, this will be a literal comma (",")
+        let commaOrNothing: String
+    }
+
+    let name: String
```
Comment (jaclync) [inline CodeGeneration/Models+Copiable.swifttemplate:80]: it took me some time to understand the `CopiableSpec` struct - this name here is the name of the class/struct that conforms to `GeneratedCopiable` right?

File: CodeGeneration/Models+Copiable.swifttemplate [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    /// "public", "private", etc.
+    let accessLevel: String
+    let properties: [Property]
+}
+
+// The collection of CopiableSpec that the template below will use. 
+let specsToGenerate: [CopiableSpec] = matchingTypes.map { type in
+    let validVariables = type.variables.filter {
+        $0.readAccess == type.accessLevel &&
```
Comment (jaclync) [inline CodeGeneration/Models+Copiable.swifttemplate:89]: just curious, what does this condition mean? like excluding this type's `private` properties?

File: CodeGeneration/Models+Copiable.swifttemplate [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            !$0.isComputed &&
+            !$0.isStatic
+    }
+    let propSpecs: [CopiableSpec.Property] = validVariables.map { variable in 
+        CopiableSpec.Property(
+            name: variable.name, 
+            typeName: variable.typeName.unwrappedTypeName,
+            copiablePropTypeName: variable.isOptional ? "NullableCopiableProp" : "CopiableProp",
+            commaOrNothing: variable == validVariables.last  ? "" : ","
```
Comment (jaclync) [inline CodeGeneration/Models+Copiable.swifttemplate:98]: ah I remember having to handle the end of class/struct/function 😅

File: CodeGeneration/Models+Copiable.swifttemplate (hunk 2 above)
Comment (jaclync) [reply CodeGeneration/Models+Copiable.swifttemplate:148]: the comments are helpful, thanks for adding them!

**Comment by shiki** [comment]
Thank you for testing again, @jaclync! And thanks everyone for the review!

**Pull request merged**